packages/sui-react-initial-props/src/withInitialProps.js

```diff
@@ -1,37 +1,112 @@
-      initialProps: {},
-      isLoading: true
-    })
-
-    useEffect(function() {
-      Page.getInitialProps({context, routeInfo})
-        .then(initialProps => {
-          setState({initialProps, isLoading: false})
-        })
-        .catch(error => {
-          setState({initialProps: {error}, isLoading: false})
-        })
-    }, [])
+    const initialPropsFromWindowRef = useRef(windowInitialProps)
+    // used to know if initialProps has been requested at least once
+    const requestedInitialPropsOnceRef = useRef(!!windowInitialProps)
+    // create routeInfo object from current props which are updated
+    const routeInfo = createRouteInfoFromProps(props)
+    // consume sui context from the context provider
+    const suiContext = useContext(SUIContext)
+    // pathName from context is outdated, so we update it from routeInfo
+    const context = {...suiContext, pathName: routeInfo.location.pathname}
+
+    const [{initialProps, isLoading}, setState] = useState(() => ({
+      initialProps: initialPropsFromWindowRef.current || {},
+      isLoading: !initialPropsFromWindowRef.current
+    }))
+
+    useEffect(() => {
+      // check if got initial props from window, because then there's no need
+      // to request them again from client
+      if (initialPropsFromWindowRef.current) {
+        initialPropsFromWindowRef.current = undefined
+      } else {
+        setState({initialProps, isLoading: true})
+
+        Page.getInitialProps({context, routeInfo})
+          .then(initialProps => {
+            setState({initialProps, isLoading: false})
+          })
+          .catch(error => {
+            setState({initialProps: {error}, isLoading: false})
+          })
+          .finally(() => {
+            if (requestedInitialPropsOnceRef.current) return
+            requestedInitialPropsOnceRef.current = true
+          })
+      }
+    }, [routeInfo.location]) // eslint-disable-line react-hooks/exhaustive-deps
+
+    const renderPage = () => (
+      <Page {...initialProps} {...props} isLoading={isLoading} />
+    )
+
+    // if the page has a `keepMounted` property and already requested
+    // initialProps once, just keep rendering the page
+    if (Page.keepMounted && requestedInitialPropsOnceRef.current) {
+      return renderPage()
+    }
 
     const renderLoading = () => {
-      // check if the page has a renderLoading method, if not, just render nothing
+      // check if the page has a `renderLoading` method, if not, just render nothing
       return Page.renderLoading
         ? Page.renderLoading({context, routeInfo})
         : null
     }
 
-    return isLoading ? renderLoading() : <Page {...initialProps} {...props} />
+    return isLoading ? renderLoading() : renderPage()
   }
+
+  // if `keepMounted` property is found and the component is the same one,
+  // we just reuse it instead of returning a new one
+  if (Page.keepMounted && Page.name === latestClientPage?.Page?.name) {
```

Comment:
Shouldn't we use displayName instead name?
```suggestion
  if (Page.keepMounted && Page.displayName === latestClientPage?.Page?.displayName) {
```